File: src/main/java/com/jore/epoc/dto/CompanySimulationStepDto.java
package com.jore.epoc.dto;

import java.time.YearMonth;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.jore.datatypes.money.Money;
import com.jore.jpa.DataTransferObject;

import lombok.Data;

@Data
/**
 * This is the main DTO for the display of company step related information
 */
public class CompanySimulationStepDto implements DataTransferObject {
    private Long id;
    private String companyName;
    private YearMonth simulationMonth;
    private CreditLineDto creditLine;
    private Money companyValue;
    private List<FactoryDto> factories = new ArrayList<>();
    private List<StorageDto> storages = new ArrayList<>();
    private List<MarketDto> markets = new ArrayList<>();
    private List<DistributionInMarketDto> distributionInMarkets = new ArrayList<>();
    private List<MessageDto> messages = new ArrayList<>();
    private List<CompanyOrderDto> orders = new ArrayList<>();

    public void addDistributionInMarket(DistributionInMarketDto distributionInMarketDto) {
        distributionInMarkets.add(distributionInMarketDto);
    }

    public void addFactory(FactoryDto factoryDto) {
        factories.add(factoryDto);
    }

    public void addMarket(MarketDto marketDto) {
        markets.add(marketDto);
    }

    public void addMessage(MessageDto messageDto) {
        messages.add(messageDto);
    }

    public void addOrder(CompanyOrderDto orderDto) {
        orders.add(orderDto);
    }

    public void addStorage(StorageDto storageDto) {
        storages.add(storageDto);
    }

    public List<MessageDto> getMessages() {
        return Collections.unmodifiableList(messages);
    }
}

File: EpocServer/Sources/App/DTOs/CompanySimulationStepDTO.swift
import Vapor

public struct CompanySimulationStepDTO: Content {
    public var id: Int64?
    public var companyName: String?
    public var simulationMonth: String?
    public var companyValue: String?
    public var factories: [FactoryDTO]
    public var storages: [StorageDTO]
    public var distributionInMarkets: [DistributionInMarketDTO]
    public var markets: [MarketDTO]
    public var messages: [MessageDTO]
    public var orders: [CompanyOrderDTO]

    public init() {
        factories = []
        storages = []
        distributionInMarkets = []
        markets = []
        messages = []
        orders = []
    }
}

public struct FactoryDTO: Content {
    public var id: Int64?
    public var capacity: Int?
}

public struct StorageDTO: Content {
    public var id: Int64?
    public var capacity: Int?
}

public struct DistributionInMarketDTO: Content {
    public var id: Int64?
}

public struct MarketDTO: Content {
    public var id: Int64?
    public var name: String?
    public var laborForce: Int?
    public var costToEnterMarket: String?
    public var distributionCost: String?
}

public struct MessageDTO: Content {
    public var level: String?
    public var message: String?
    public var relevantMonth: String?
}

public struct CompanyOrderDTO: Content {
    public var orderType: String?
    public var amount: String?
}
